python/user/bot/user_bot.py

```diff
@@ -1,20 +1,37 @@
 import datetime
 import logging
 import os
-
+import asyncio
 import dazl
-from dazl import exercise
+import math
+from dazl import exercise, ContractId
+import heapq
+from dataclasses import dataclass
 
 dazl.setup_default_logger(logging.INFO)
 
 EPOCH = datetime.datetime.utcfromtimestamp(0)
 
 
 class Chat:
-    ArchiveMessagesRequest = 'Chat.V2:ArchiveMessagesRequest'
-    Message = 'Chat.V2:Message'
-    UserSettings = 'Chat.V2:UserSettings'
+    ArchiveMessagesRequest = 'Chat.V3:ArchiveMessagesRequest'
+    Message = 'Chat.V3:Message'
+    UserSettings = 'Chat.V3:UserSettings'
+
+
+@dataclass(frozen=True)
+class Message:
+    post_at: int
+    cid: 'ContractId'
 
+    def __lt__(self, other):
+        return self.post_at < other.post_at
+
+
+# default to 14 days as defined in DAML
+archive_after = 1209600
+message_heap: '[Message]' = []
```

Comment:
Could this be moved into the `main` function instead at least? Boo globals…

Reply:
hmmm, basically I want it to be static, if I move it into main,  the eventHandler needs to reference it via `nonlocal`?

Reply:
or it's better to wrap the whole bot in an object, to avoid the var shadowing?